evalml/model_understanding/decision_boundary.py

```diff
@@ -0,0 +1,234 @@
+    sens = _recall(val_list)
+    if val_list[1] == 0:
+        spec = 0
+    else:
+        spec = val_list[1] / (val_list[1] + val_list[2])
+    return (sens + spec) / 2
+
+
+def _precision(val_list):
+    """Helper function to help us find the precision.
+
+    Args:
+        val_list (list): The confusion matrix input, expected format is [tp, tn, fp, fn].
+
+    Returns:
+        float: Precision.
+    """
+    if val_list[0] == 0:
+        return 0
+    return val_list[0] / (val_list[0] + val_list[2])
+
+
+def _recall(val_list):
+    """Helper function to help us find the recall.
+
+    Args:
+        val_list (list): The confusion matrix input, expected format is [tp, tn, fp, fn].
+
+    Returns:
+        float: Recall.
+    """
+    if val_list[0] == 0:
+        return 0
+    return val_list[0] / (val_list[0] + val_list[3])
+
+
+def _f1(val_list):
+    """Helper function to help us find the F1 score.
+
+    Args:
+        val_list (list): The confusion matrix input, expected format is [tp, tn, fp, fn].
+
+    Returns:
+        float: F1 Score.
+    """
+    prec = _precision(val_list)
+    rec = _recall(val_list)
+    if prec * rec == 0:
+        return 0
+    return 2 * (prec * rec) / (prec + rec)
+
+
+def _find_confusion_matrix_objective_threshold(pos_skew, neg_skew, ranges):
+    """Iterates through the arrays and determines the ideal objective thresholds and confusion matrix values for each threshold value.
+
+    Args:
+        pos_skew (list): The number of rows per bin value for the actual postive values.
+        neg_skew (list): The number of rows per bin value for the actual negative values.
+        ranges (list): The bin ranges, spanning from 0.0 to 1.0. The length of this list - 1 is equal to the number of bins.
+
+    Returns:
+        tuple: The first element is a list of confusion matrix values at each threshold bin, and the second element
+            is a dictionary with the ideal objective thresholds and associated objective scores.
+    """
+    thresh_conf_matrix_list = []
+    num_fn, num_tn = 0, 0
+    total_pos, total_neg = sum(pos_skew), sum(neg_skew)
+    objective_dict = {
+        "accuracy": [{"objective score": 0, "threshold value": 0}, _accuracy],
+        "balanced_accuracy": [
+            {"objective score": 0, "threshold value": 0},
+            _balanced_accuracy,
+        ],
+        "precision": [{"objective score": 0, "threshold value": 0}, _precision],
+        "f1": [{"objective score": 0, "threshold value": 0}, _f1],
+    }
+    for i, thresh_val in enumerate(ranges[1:]):
+        num_fn += pos_skew[i]
+        num_tn += neg_skew[i]
+        num_tp = total_pos - num_fn
+        num_fp = total_neg - num_tn
+        # this is also the confusion matrix
+        val_list = [num_tp, num_tn, num_fp, num_fn]
+        thresh_conf_matrix_list.append(val_list)
+
+        # let's iterate through the list to find the vals
+        for k, v in objective_dict.items():
+            obj_val = v[1](val_list)
```

Comment:
I kept this call rather than moving towards using the input `y/y_pred` because it's much faster on larger datasets/larger `n_bins` (see the PR description for a time comparison!)